File: cpp/TokenUtil.cpp
#include "TokenUtil.h"
#include <iostream>
#include <intrin.h>

static bitboard calcWinCheckPattern(long index)
{
	bitboard pattern = TokenUtil::EMPTY;
	for (long y = 0; y < TokenUtil::HEIGHT; y++)
	{
		for (long x = 0; x < TokenUtil::WIDTH; x++)
		{
			if (index == ((x + 2 * y) & 3))
			{
				pattern |= TokenUtil::nthBit(TokenUtil::position(x, y));
			}
		}
	}
	return pattern;
}
const bitboard WIN_CHECK_PATTERNS[] = { calcWinCheckPattern(0), calcWinCheckPattern(1), calcWinCheckPattern(2), calcWinCheckPattern(3) };

bitboard TokenUtil::lowestOneBit(bitboard board)
{
	return board & (~board + 1);
}

bitboard TokenUtil::clearLowestOneBit(bitboard board)
{
	return board & (board - 1);
}

unsigned long TokenUtil::bitCount(bitboard board)
{
	return (unsigned long)__popcnt64(board);
}

unsigned long TokenUtil::trailingZeros(bitboard board)
{
	return (unsigned long)_tzcnt_u64(board);
}

bitboard TokenUtil::nthBit(unsigned long position)
{
	return FIRST_BIT << position;
}

bitboard TokenUtil::nBits(unsigned long count)
{
	return nthBit(count) - FIRST_BIT;
}

bool TokenUtil::isSymmetrical(bitboard board)
{
	for (long x = 0; x < WIDTH / 2; x++)
	{
		bitboard left = board >> (x * RIGHT);
		bitboard right = board >> ((WIDTH - x - 1) * RIGHT);
		if (((left ^ right) & COLUMN_0))
		{
			return false;
		}
	}
	return true;
}

bitboard TokenUtil::mirror(bitboard board)
{
	bitboard result = board & CENTER_COLUMN;
	for (long x = 0; x < WIDTH / 2; x++)
	{
		long mirrorX = WIDTH - x - 1;
		long offset = mirrorX - x;

		result |= (board & (COLUMN_0 << (x * RIGHT))) << (offset * RIGHT);
		result |= (board & (COLUMN_0 << (mirrorX * RIGHT))) >> (offset * RIGHT);
	}
	return result;
}

bool TokenUtil::isGameOver(bitboard own, bitboard opp)
{
	return isBoardFull(own, opp) || isWin(opp);
}

bool TokenUtil::isBoardFull(bitboard own, bitboard opp)
{
	return occupied(own, opp) == FULL_BOARD;
}

bool TokenUtil::isWin(bitboard board)
{
	return isNonVerticalWin(board)
		|| squish(board, UP);
}

bool TokenUtil::isNonVerticalWin(bitboard board)
{
	return squish(board, RIGHT_UP)
		|| squish(board, RIGHT_DOWN)
		|| squish(board, RIGHT);
}

bool TokenUtil::canWin(bitboard board, bitboard moves)
{
	for (bitboard pattern : WIN_CHECK_PATTERNS)
	{
		bitboard patternMoves = moves & pattern;
		if (patternMoves && isWin(move(board, patternMoves)))
		{
			return true;
		}
	}
	return false;
}

bitboard TokenUtil::move(bitboard board, bitboard move)
{
	return board | move;
}

bitboard TokenUtil::findAnyWinningMove(bitboard board, bitboard moves)
{
	bitboard squished = squish(move(board, moves), UP);
	if (squished)
	{
		return lowestOneBit(stretch(squished, UP) & moves);
	}
	for (bitboard pattern : WIN_CHECK_PATTERNS)
	{
		bitboard patternMoves = moves & pattern;
		if (patternMoves)
		{
			bitboard moved = move(board, patternMoves);
			squished = squish(moved, RIGHT);
			if (squished)
			{
				return lowestOneBit(stretch(squished, RIGHT) & moves);
			}

			squished = squish(moved, RIGHT_DOWN);
			if (squished)
			{
				return lowestOneBit(stretch(squished, RIGHT_DOWN) & moves);
			}

			squished = squish(moved, RIGHT_UP);
			if (squished)
			{
				return lowestOneBit(stretch(squished, RIGHT_UP) & moves);
			}
		}
	}
	return EMPTY;
}

bitboard TokenUtil::threats(bitboard own, bitboard opp)
{
	bitboard full = occupied(own, opp);
	bitboard free = unoccupied(full);
	bitboard squishedRight = EMPTY, squishedRightDown = EMPTY, squishedRightUp = EMPTY;
	for (bitboard pattern : WIN_CHECK_PATTERNS)
	{
		bitboard ownPattern = move(own, free & pattern);
		squishedRight |= squish(ownPattern, RIGHT);
		squishedRightDown |= squish(ownPattern, RIGHT_DOWN);
		squishedRightUp |= squish(ownPattern, RIGHT_UP);
	}
	bitboard squishedUp = squish(move(own, moves(own, opp)), UP);
	return (stretch(squishedUp, UP)
		| stretch(squishedRight, RIGHT)
		| stretch(squishedRightDown, RIGHT_DOWN)
		| stretch(squishedRightUp, RIGHT_UP))
		& free;
}

bitboard TokenUtil::squish(bitboard board, unsigned long direction)
{
	board &= board << (2 * direction);
	board &= board << direction;
	return board;
}

bitboard TokenUtil::stretch(bitboard board, unsigned long direction)
{
	board |= board >> (2 * direction);
	board |= board >> direction;
	return board;
}

bitboard TokenUtil::moves(bitboard own, bitboard opp)
{
	return  (occupied(own, opp) + ROW_0) & FULL_BOARD;
}

bitboard TokenUtil::occupied(bitboard own, bitboard opp)
{
	return own | opp;
}

bitboard TokenUtil::unoccupied(bitboard occupied)
{
	return FULL_BOARD ^ occupied;
}

bitboard TokenUtil::hash(bitboard own, bitboard opp)
{
	return GOLDEN_MULTIPLIER * id(own, opp);
}

bitboard TokenUtil::id(bitboard own, bitboard opp)
{
	return (occupied(own, opp) + ROW_0) ^ own;
}

unsigned long TokenUtil::position(unsigned long x, unsigned long y)
{
	return x * BUFFERED_HEIGHT + y;
}

std::string TokenUtil::toString(bitboard own, bitboard opp)
{
	std::string str = "";
	for (long y = HEIGHT - 1; y >= 0; y--)
	{
		for (unsigned long x = 0; x < WIDTH; x++)
		{
			bitboard bit = nthBit(position(x, y));
			if (bit & own)
			{
				str.append("[x]");
			}
			else if (bit & opp)
			{
				str.append("[o]");
			}
			else
			{
				str.append("[ ]");
			}
		}
		if (y)
		{
			str.append("\n");
		}
	}
	return str;
}

std::string TokenUtil::toString(bitboard board)
{
	return toString(board, EMPTY);
}

File: java/src/main/java/com/etherblood/connect4/sandbox/TokenUtil.java
package com.etherblood.connect4.sandbox;

import com.etherblood.connect4.Util;

public class TokenUtil {

    public static final int WIDTH, HEIGHT, BUFFERED_HEIGHT;
    public static final long ROW_0, COLUMN_0, BUFFERED_COLUMN_0;
    public static final long FULL_BOARD, LEFT_SIDE, CENTER_COLUMN;
    public static final long EVEN_INDEX_ROWS, ODD_INDEX_ROWS;
    public static final int RIGHT, RIGHT_DOWN, RIGHT_UP, UP;
    public static final boolean IS_HEIGHT_EVEN;

    public static final long[] WIN_CHECK_PATTERNS = new long[4];

    static {
        WIDTH = 7;
        HEIGHT = 6;

        //computed settings
        BUFFERED_HEIGHT = HEIGHT + 1;
        if (WIDTH * BUFFERED_HEIGHT > Long.SIZE) {
            throw new IllegalArgumentException();
        }
        IS_HEIGHT_EVEN = (HEIGHT & 1) == 0;
        COLUMN_0 = Util.toLongMask(HEIGHT);
        BUFFERED_COLUMN_0 = Util.toLongMask(BUFFERED_HEIGHT);
        ROW_0 = Util.toLongMask(WIDTH * BUFFERED_HEIGHT) / Util.toLongMask(BUFFERED_HEIGHT);
        UP = 1;
        RIGHT = BUFFERED_HEIGHT;
        RIGHT_UP = RIGHT + UP;
        RIGHT_DOWN = RIGHT - UP;
        FULL_BOARD = ROW_0 * COLUMN_0;
        LEFT_SIDE = FULL_BOARD >>> (Util.ceilDiv(WIDTH, 2) * RIGHT);
        CENTER_COLUMN = (WIDTH & 1) != 0 ? COLUMN_0 << (WIDTH / 2 * RIGHT) : 0;

        EVEN_INDEX_ROWS = (COLUMN_0 / 3) * ROW_0;
        ODD_INDEX_ROWS = FULL_BOARD ^ EVEN_INDEX_ROWS;

        for (int y = 0; y < HEIGHT; y++) {
            for (int x = 0; x < WIDTH; x++) {
                WIN_CHECK_PATTERNS[(x + 2 * y) & 3] |= Util.toLongFlag(index(x, y));
            }
        }
    }

    public static long move(long tokens, long moves) {
        return tokens | moves;
    }

    public static boolean isGameOver(long ownTokens, long opponentTokens) {
        return isBoardFull(ownTokens, opponentTokens) || isWin(opponentTokens);
    }

    public static boolean isBoardFull(long ownTokens, long opponentTokens) {
        return occupied(ownTokens, opponentTokens) == FULL_BOARD;
    }

    public static boolean isWin(long tokens) {
        return isNonVerticalWin(tokens) || squish(tokens, UP) != 0;
    }

    public static boolean isNonVerticalWin(long tokens) {
        return squish(tokens, RIGHT) != 0
                || squish(tokens, RIGHT_DOWN) != 0
                || squish(tokens, RIGHT_UP) != 0;
    }

    public static boolean canWin(long ownTokens, long opponentTokens) {
        return (threats(ownTokens, opponentTokens) & generateMoves(ownTokens, opponentTokens)) != 0;
    }

    public static long threats(long ownTokens, long opponentTokens) {
        long free = unoccupied(ownTokens, opponentTokens);
        long squishedRight = 0, squishedRightDown = 0, squishedRightUp = 0;
        for (long pattern : WIN_CHECK_PATTERNS) {
            long ownPattern = move(ownTokens, free & pattern);
            squishedRight |= squish(ownPattern, RIGHT);
            squishedRightDown |= squish(ownPattern, RIGHT_DOWN);
            squishedRightUp |= squish(ownPattern, RIGHT_UP);
        }
        long squishedUp = squish(move(ownTokens, generateMoves(ownTokens, opponentTokens)), UP);
        return (stretch(squishedUp, UP)
                | stretch(squishedRight, RIGHT)
                | stretch(squishedRightDown, RIGHT_DOWN)
                | stretch(squishedRightUp, RIGHT_UP))
                & free;
    }

    public static long squish(long tokens, int directionShift) {
        tokens &= tokens >>> (2 * directionShift);
        tokens &= tokens >>> directionShift;
        return tokens;
    }

    public static long stretch(long tokens, int directionShift) {
        tokens |= tokens << (2 * directionShift);
        tokens |= tokens << directionShift;
        return tokens;
    }

    public static long generateMoves(long ownTokens, long opponentTokens) {
        return (occupied(ownTokens, opponentTokens) + ROW_0) & FULL_BOARD;
    }

    public static long occupied(long ownTokens, long opponentTokens) {
        return ownTokens | opponentTokens;
    }

    public static long unoccupied(long ownTokens, long opponentTokens) {
        return FULL_BOARD ^ occupied(ownTokens, opponentTokens);
    }

    public static long upperBoundId() {
        return Util.toLongMask(WIDTH * BUFFERED_HEIGHT);
    }

    public static long id(long ownTokens, long opponentTokens) {
        return (occupied(ownTokens, opponentTokens) + ROW_0) ^ ownTokens;
    }

    public static String toString(long mask) {
        return toString(mask, 0);
    }

    public static String toString(long ownTokens, long opponentTokens) {
        return toString(ownTokens, opponentTokens, 0);
    }

    public static String toString(long ownTokens, long opponentTokens, long threats) {
        if ((ownTokens | opponentTokens | FULL_BOARD) != FULL_BOARD) {
            throw new IllegalArgumentException();
        }
        if ((ownTokens & opponentTokens) != 0) {
            throw new IllegalArgumentException();
        }
        StringBuilder builder = new StringBuilder();
        for (int y = HEIGHT - 1; y >= 0; y--) {
            for (int x = 0; x < WIDTH; x++) {
                long flag = Util.toLongFlag(index(x, y));
                if((threats & flag) != 0) {
                    builder.append('(');
                } else {
                    builder.append('[');
                }
                if ((flag & ownTokens) != 0) {
                    builder.append('x');
                } else if ((flag & opponentTokens) != 0) {
                    builder.append('o');
                } else {
                    builder.append(' ');
                }
                if((threats & flag) != 0) {
                    builder.append(')');
                } else {
                    builder.append(']');
                }
            }
            if (y != 0) {
                builder.append('\n');
            }
        }
        return builder.toString();
    }

    public static String toString(int[] board, int offset) {
        StringBuilder builder = new StringBuilder();
        for (int y = HEIGHT - 1; y >= 0; y--) {
            for (int x = 0; x < WIDTH; x++) {
                builder.append("[");
                builder.append(board[offset + index(x, y)]);
                builder.append("]");
            }
            if (y != 0) {
                builder.append('\n');
            }
        }
        return builder.toString();
    }

    public static int index(int x, int y) {
        return x * RIGHT + y;
    }
}
